compiler/noirc_frontend/src/elaborator/trait_impls.rs

```diff
@@ -0,0 +1,351 @@
+            let override_constraint_is_from_impl =
+                trait_impl_where_clause.iter().any(|impl_constraint| {
+                    impl_constraint.trait_id == override_trait_constraint.trait_id
+                });
+            if override_constraint_is_from_impl {
+                continue;
+            }
+
+            let substituted_constraint_type = Self::reset_generics_on_constraint_type(
+                &override_trait_constraint.typ,
+                &override_generics,
+                &method_generics,
+                trait_impl_generics_len,
+                trait_generics_len,
+            );
+
+            if !substituted_method_ids.contains(&(
+                substituted_constraint_type.clone(),
+                override_trait_constraint.trait_id,
+            )) {
+                let the_trait = self.interner.get_trait(override_trait_constraint.trait_id);
+                self.push_err(DefCollectorErrorKind::ImplIsStricterThanTrait {
+                    constraint_typ: override_trait_constraint.typ,
+                    constraint_name: the_trait.name.0.contents.clone(),
+                    constraint_span: override_trait_constraint.span,
+                    trait_method_name: method.name.0.contents.clone(),
+                    trait_method_span: method.location.span,
+                });
+            }
+        }
+    }
+
+    /// Resets the generics of a trait impl's trait constraint type.
+    /// This helps us match the trait constraint type from a trait itself and its trait impl.
+    /// If a type contains named generics, this method will reset the type variable ids
+    /// of those generics based off its position in the actual trait definition's generics.
+    ///
+    /// Example
+    ///
+    /// ```
+    /// trait MyTrait { }
+    /// trait Foo<T> {
+    ///     fn foo_good<U>() where T: MyTrait;
+    ///
+    ///     fn foo_bad<U>() where T: MyTrait;
+    /// }
+    /// impl<A> Foo<A> for () {
+    ///    // Error issued here as `foo` does not have the `MyTrait` constraint
+    ///    fn foo_good<B>() where A: MyTrait {}
+    ///
+    ///     fn foo_bad<B>() where B: MyTrait {}
+    /// ```
+    /// `A` in `A: MyTrait` will have an actual type variable id based upon the global interner current type variable id.
+    /// In the example, above we will have A -> '2. However, we want to compare again T -> '0.
+    /// We can find `A`'s position among the trait generics and reset it to be T -> '0.
+    ///  
+    /// On the flip side, `B` in `B: MyTrait` will be reset to T -> '1.
+    /// This will not match T -> '0 and we can mark that these types are unequal and our impl is stricter than the trait.
+    ///
+    /// The last two fields `trait_impl_generics_len` and `trait_generics_len` are only necessary to account
+    /// for extra impl generics when indexing into the trait definition's generics.
+    fn reset_generics_on_constraint_type(
```

Comment:
It sounds like we should be using `substitute` instead to substitute the impl's generics for the trait's or vice-versa.

Reply:
E.g. check out `check_trait_impl_method_matches_declaration`

Reply:
`instantiate_with_bindings` only substitutes with `Type::Forall`? This method does use `substitute` internally. 

I will at using the method in `check_trait_impl_method_matches_declaration` though to substitute the generics on the trait function with that of the impl function.

Reply:
Right, so you'll probably want to use `substitute` in this case

Reply:
I have submitted an update that uses `substitute` on the type itself. I see that this method I made was overly complex and redoing a lot of the logic that is handled by `substitute` itself.